bob/src/api/s3.rs

```diff
@@ -0,0 +1,266 @@
+        if time > last_modified {
+            return Err(StatusS3::Status(StatusCode::NOT_MODIFIED));
+        }
+    }
+    if let Some(time) = headers.if_unmodified_since {
+        if time < last_modified {
+            return Err(StatusS3::Status(StatusCode::PRECONDITION_FAILED));
+        }
+    }
+    Ok(GetObjectOutput { data, content_type })
+}
+
+// PUT /s3/default/:key
+async fn put_object<A>(
+    Extension(bob): Extension<&BobServer<A>>,
+    Path(key): Path<String>,
+    body: Bytes,
+    headers: CopyObjectHeaders,
+    creds: CredentialsHolder<A>,
+) -> Result<StatusS3, StatusS3>
+where
+    A: Authenticator,
+{
+    if !bob.auth().check_credentials(creds.into_credentials())?.has_write() {
+        return Err(AuthError::PermissionDenied.into());
+    }
+    let key = DataKey::from_str(&key)?.0;
+    if headers.is_source_key_set() {
+        return copy_object(bob, key, headers).await;
+    }
+    let data_buf = body.to_vec();
+    let data = BobData::new(
+        data_buf,
+        BobMeta::new(chrono::Local::now().timestamp() as u64),
+    );
+
+    let opts = BobOptions::new_put(None);
+    bob.grinder().put(key, data, opts).await?;
+
+    Ok(StatusS3::from(StatusExt::from(StatusCode::CREATED)))
+}
+
+#[derive(Debug)]
+pub(crate) struct CopyObjectHeaders {
+    if_modified_since: Option<u64>,
+    if_unmodified_since: Option<u64>,
+    source_key: Option<DataKey>,
+}
+
+impl CopyObjectHeaders {
+    fn is_source_key_set(&self) -> bool {
+        self.source_key.is_some()
+    }
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for CopyObjectHeaders
+where
+    B: Send,
+{
+    type Rejection = StatusS3;
+    async fn from_request(request: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let headers = request
+            .headers()
+            .expect("headers removed by another extractor");
+        let source_key = headers.get("x-amz-copy-source").and_then(|x| {
+            let key = x.to_str().map(|s| s.to_string()).ok()?;
+            DataKey::from_str(&key).ok()
+        });
+        Ok(CopyObjectHeaders {
+            if_modified_since: headers
+                .get("If-Modified-Since")
+                .and_then(|x| {
+                    let s = x.to_str().expect("failed to convert header to str");
+                    chrono::DateTime::parse_from_rfc2822(s).ok()
+                })
+                .and_then(|x| x.timestamp().try_into().ok()),
+            if_unmodified_since: headers
+                .get("If-Unmodified-Since")
+                .and_then(|x| {
+                    let s = x.to_str().expect("failed to convert header to str");
+                    chrono::DateTime::parse_from_rfc2822(s).ok()
+                })
+                .and_then(|x| x.timestamp().try_into().ok()),
+            source_key,
+        })
+    }
+}
+
+async fn copy_object<A: Authenticator>(
+    bob: &BobServer<A>,
+    key: BobKey,
+    headers: CopyObjectHeaders,
+) -> Result<StatusS3, StatusS3> {
```

Comment:
Authorization is missing

Reply:
This function is called in PUT after authorization is done